aptos-move/framework/aptos-framework/sources/configs/stake.move

```diff
@@ -860,24 +864,33 @@ module aptos_framework::stake {
 
     /// Update the number of missed votes. This is only called by block::prologue().
     /// This function cannot abort.
-    public(friend) fun update_performance_statistics(missed_votes: vector<u64>) acquires ValidatorPerformance {
+    public(friend) fun update_performance_statistics(proposer_index_optional: vector<u64>, failed_proposer_indices: vector<u64>) acquires ValidatorPerformance {
         // Validator set cannot change until the end of the epoch, so the validator index in list of missed votes should
         // match with those of the missed vote counts in ValidatorPerformance resource.
         let validator_perf = borrow_global_mut<ValidatorPerformance>(@aptos_framework);
-        let validator_missed_votes_counts = &mut validator_perf.missed_votes;
-        let i = 0;
-        let len = vector::length(&missed_votes);
-        while (i < len) {
-            let validator_index = *vector::borrow(&missed_votes, i);
-            // Skip any validator indices that are out of bounds, this ensures that this function doesn't abort if there
-            // are out of bounds errors.
-            if (validator_index < vector::length(validator_missed_votes_counts)) {
-                let missed_votes_count = vector::borrow_mut(validator_missed_votes_counts, validator_index);
-                *missed_votes_count = *missed_votes_count + 1;
+        let validator_len = vector::length(&validator_perf.validators);
+
+        // proposer_indices is a vector because it can be missing (for NilBlocks)
+        if (vector::length(&proposer_index_optional) >= 1) {
+            let proposer_index = *vector::borrow(&proposer_index_optional, 0);
+            // Here, and in all other vector::borrow, skip any validator indices that are out of bounds,
+            // this ensures that this function doesn't abort if there are out of bounds errors.
+            if (proposer_index < validator_len) {
+                let validator = vector::borrow_mut(&mut validator_perf.validators, proposer_index);
+                validator.successful_proposals = validator.successful_proposals + 1;
             };
-            i = i + 1;
         };
-        validator_perf.num_blocks = validator_perf.num_blocks + 1;
+
+        let f = 0;
+        let f_len = vector::length(&failed_proposer_indices);
+        while (f < f_len) {
```

Comment:
Hmm..orthogonal to this but `Move` could really use `for loop` support - writing these `while` loops seems tedious and unclean :)

Reply:
+1000. We might not have time to make it before mainnet though